Userland/Applications/PackageManager/main.cpp

```diff
@@ -0,0 +1,186 @@
+    DeprecatedString method = "GET";
+    outln("pkg: Syncing packages database...");
+    request = protocol_client->start_request(method, url, request_headers, ReadonlyBytes {}, proxy_data);
+    request->on_finish = [&](bool success, auto) {
+        if (!success)
+            outln("pkg: Syncing packages database failed.");
+        else
+            outln("pkg: Syncing packages database done.");
+        loop.quit(success ? 0 : 1);
+    };
+    request->stream_into(*output_stream);
+    return loop.exec();
+}
+
+class Port {
+public:
+    enum class Type {
+        Auto,
+        Dependency,
+        Manual,
+    };
+
+    Port(String name, Type type, String version)
+        : m_name(move(name))
+        , m_type(type)
+        , m_version(move(version))
+    {
+    }
+
+    Type type() const { return m_type; }
+    StringView type_as_string_view() const
+    {
+        if (m_type == Type::Auto)
+            return "Automatic"sv;
+        if (m_type == Type::Dependency)
+            return "Dependency"sv;
+        if (m_type == Type::Manual)
+            return "Manual"sv;
+        VERIFY_NOT_REACHED();
+    }
+
+    StringView name() const { return m_name.bytes_as_string_view(); }
+    StringView version() const { return m_version.bytes_as_string_view(); }
+
+private:
+    String m_name;
+    Type m_type;
+    String m_version;
+};
+
+static ErrorOr<void> print_installed_ports(Optional<Port::Type> filter_type, Function<void(Port const&)> callback)
+{
+    Vector<Port> ports;
+    auto input_stream = TRY(Core::File::open("/usr/Ports/installed.db"sv, Core::File::OpenMode::Read));
+    auto whole_file = TRY(input_stream->read_until_eof());
+
+    size_t pointer = 0;
+    while (pointer < whole_file.size()) {
+        StringView current_view = ReadonlyBytes(whole_file.bytes().slice(pointer));
+        auto newline_separator_index = current_view.find('\n');
+        VERIFY(newline_separator_index.has_value());
+        current_view = current_view.substring_view(0, newline_separator_index.value());
+        if (current_view.starts_with("dependency"sv)) {
+            auto parts = current_view.split_view(' ');
+            auto type = Port::Type::Dependency;
+            // FIXME: Add versioning when printing these ports!
+            ports.append(Port { TRY(String::from_utf8(parts[1])), type, TRY(String::from_utf8(""sv)) });
+        } else {
+            auto parts = current_view.split_view(' ');
+            VERIFY(parts.size() == 3);
+            auto type = Port::Type::Auto;
+            if (parts[0] == "manual"sv) {
+                type = Port::Type::Manual;
+            } else if (parts[0] == "auto"sv) {
+                type = Port::Type::Auto;
+            } else {
+                VERIFY_NOT_REACHED();
+            }
+            ports.append(Port { TRY(String::from_utf8(parts[1])), type, TRY(String::from_utf8(parts[2])) });
+        }
+        pointer += current_view.length() + 1;
+    };
```

Comment:
it seems you could extract that into something called e.g. `collect_installed_ports`; iterating and filtering feels like completely separate problem. (youʼre doing that possibly twice, depending on api/features you want.)